### PR TITLE
Nullable: System.Runtime.CompilerServices

### PR DESCRIPTION
**Pull request by stephentoub**

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs [hunk 1]
```diff
@@ -894,7 +895,7 @@ internal static Task<TResult> GetTaskForResult(TResult result)
                 // For Boolean, we cache all possible values.
                 if (typeof(TResult) == typeof(bool)) // only the relevant branches are kept for each value-type generic instantiation
                 {
-                    bool value = (bool)(object)result;
+                    bool value = (bool)(object?)result;
```
Comment (safern) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs]: Just to learn, why are this double casts needed? why not directly into `bool`?

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/DateTimeConstantAttribute.cs [hunk 2]
```diff
@@ -14,6 +15,6 @@ public DateTimeConstantAttribute(long ticks)
             _date = new DateTime(ticks);
         }
 
-        public override object Value => _date;
+        public override object? Value => _date; // TODO-NULLABLE: https://github.com/dotnet/roslyn/issues/23268
```
Comment (safern) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/DateTimeConstantAttribute.cs]: We're already supressing the warning you're referring to in the issue, why is this needed?

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/YieldAwaitable.cs [hunk 3]
```diff
@@ -188,7 +187,7 @@ private static Action OutputCorrelationEtwEvent(Action continuation)
             private static readonly SendOrPostCallback s_sendOrPostCallbackRunAction = RunAction;
             /// <summary>Runs an Action delegate provided as state.</summary>
             /// <param name="state">The Action delegate to invoke.</param>
-            private static void RunAction(object state) { ((Action)state)(); }
+            private static void RunAction(object? state) { ((Action)state!)(); }
```
Comment (safern) [inline src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/YieldAwaitable.cs]: Why accept null here and then use `!`?

**Review by safern — APPROVED** [review]
Left few comments but LGTM

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs]: The compiler won't let you cast directly from `T` to `bool` as it doesn't know what `T` is and whether such a cast actually exists.  By first going to `object`, we leave it up to the runtime, effectively boxing the `T` (which the compiler allows) and then unboxing it (which the compiler allows, leaving verification up to the runtime).  The JIT then has an optimization to remove the boxing for this specific pattern.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/YieldAwaitable.cs (hunk 3 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/YieldAwaitable.cs]: `RunAction` is being used with a delegate that takes an `Action<object?>` because that delegate is passed to a method that expects an `Action<object?>`.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/DateTimeConstantAttribute.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/DateTimeConstantAttribute.cs]: Because without this we get a warning:
```
shared\System\Runtime\CompilerServices\DateTimeConstantAttribute.cs(18,32): error CS8608: Nullability of reference types in type doesn't match overridden member. [d:\repos\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
```
As @cston commented at https://github.com/dotnet/corert/pull/7276#issuecomment-481717253, "`#nullable enable` will re-enable any nullable warnings that were disabled in the project file".

Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/DateTimeConstantAttribute.cs]: 😢 ok makes sense.

File: src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/YieldAwaitable.cs (hunk 3 above)
Comment (safern) [reply src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/YieldAwaitable.cs]: Thanks for elaborating.
